### PR TITLE
fix(payment): PAYPAL-1635 Buy Now feature for amazon pay

### PR DESCRIPTION
**Pull request by bc-nick**
## What?

Added Buy Now functionality for amazon pay on the Product Details Page

## Why?

According to task [PAYPAL-1635](https://bigcommercecloud.atlassian.net/browse/PAYPAL-1635)

## Testing / Proof

Testing with dependent PR

https://user-images.githubusercontent.com/99336044/210571152-d0df58f2-190a-4c6f-9967-66ab3fef6cf4.mov

Testing the rest of the Amazon flow to make sure that everything works correctly with our changes

- Amazon button flow

https://user-images.githubusercontent.com/99336044/211554560-ed067fbc-6033-4015-a0cf-ce40351b85db.mov

https://user-images.githubusercontent.com/99336044/211554616-9b0ade8b-7c25-46a1-a77e-906874036877.mov

- Amazon payment flow

https://user-images.githubusercontent.com/99336044/211575527-22ee341d-efc9-462b-a9ed-a6f326661e15.mov

https://user-images.githubusercontent.com/99336044/211575649-a39733af-45a0-4e6e-b47f-7406c9925eeb.mov

## Depends on 

https://github.com/bigcommerce/bigcommerce/pull/50031

@bigcommerce/checkout @bigcommerce/payments

[PAYPAL-1635]: https://bigcommercecloud.atlassian.net/browse/PAYPAL-1635?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PAYPAL-1635]: https://bigcommercecloud.atlassian.net/browse/PAYPAL-1635?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by mauricio-sg** [comment]
@bc-nick I sent you a DM

File: packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts:174]: I'm not sure if Amazon's script accepts modifying this value. I've found nothing about it in their docs.

File: packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts:220]: You should set this value in your if statement instead, because not all flows require it.

File: packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2.ts
Comment (mauricio-sg) [inline packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2.ts:61]: Not sure about making all props optional because, according to Amazon docs, they are not :/

**Comment by mauricio-sg** [comment]
@bc-nick I'm good with your changes. However, I'm a bit worried because `amazon-pay-v2-payment-processor.ts` was touched and this file is the core logic for all of our various flows. I'm wondering if you performed a full regression to make sure everything else works as expected. Also, wondering if these changes need to be under an experiment. I'd say I'll approve if you could provide more evidence that everything works fine besides your changes. Anyway, at the end of the day, the final decision is made by the @bigcommerce/checkout team. Please, let me know any questions. cc @vmparra

File: packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2.ts
Comment (bc-nick) [reply packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2.ts:61]: @mauricio-sg when we initialize the strategy for **Buy Now** flow we don't have options and we collect the amazon config via this `_getAmazonPayV2ButtonOptions` method

You can see it there [L129](https://github.com/bigcommerce/checkout-sdk-js/blob/562ab380046cec191067d5c68e8f2dde221aab5c/packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts#L129)

We only have `options.buyNowInitializeOptions` witch we get from a script is loaded from bcapp
I can check if we can pass required parameters for Amazon from there but i think it is redundant because we get all these variables in `_getAmazonPayV2ButtonOptions` method

File: packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts
Comment (bc-nick) [reply packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts:174]: @mauricio-sg we can change `productType` and Amazon allows to do this.
There is an example you can see: [existing_buyer_flow](https://amazonpay-onboarding.com/ee/flow/existing_buyer_flow/flow3/)

**Comment by bc-nick** [comment]
> @bc-nick I'm good with your changes. However, I'm a bit worried because `amazon-pay-v2-payment-processor.ts` was touched and this file is the core logic for all of our various flows. I'm wondering if you performed a full regression to make sure everything else works as expected. Also, wondering if these changes need to be under an experiment. I'd say I'll approve if you could provide more evidence that everything works fine besides your changes. Anyway, at the end of the day, the final decision is made by the @bigcommerce/checkout team. Please, let me know any questions. cc @vmparra

We didn't go through a full regression. We want to do it after review.
I think we can make a full regression now and i will back to you with the results.

File: packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts
Comment (mauricio-sg) [reply packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts:174]: Cool, I didn't know that. I have to mention that I didn't know the existence of that page and, the same page says in its introduction that _"If there are any discrepancies between description in this document and Integration Guide, the [Integration Guide](https://developer.amazon.com/docs/amazon-pay-checkout/introduction.html) shall prevail."_ which is the one that Amazon provided us with to work. When I said, _"I've found nothing about it in their docs."_ I was referring to that guide, sorry for not previously posting the link. So, the button is working as we can see in the videos but it might be worth double-checking with the Amazon team if it's really fully supported.

File: packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2.ts
Comment (mauricio-sg) [reply packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2.ts:61]: Makes sense. How about modifying instead `packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts` and doing something like this:

```js
interface WithBuyNowFeatures {
    /**
     * The options that are required to initialize Buy Now functionality.
     */
    buyNowInitializeOptions?: {
        getBuyNowCartRequestBody?(): BuyNowCartRequestBody | void;
    };
}

/**
 * The required config to render the AmazonPayV2 buttton.
 */
export type AmazonPayV2ButtonInitializeOptions = AmazonPayV2ButtonParameters | WithBuyNowFeatures;
```

File: packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts
Comment (animesh1987) [reply packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2-payment-processor.ts:174]: Hey @bc-nick can you check the comment Mauricio posted as well and provide testing screenshot for the same as well. Thanks.

File: packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2.ts
Comment (bc-nick) [reply packages/core/src/payment/strategies/amazon-pay-v2/amazon-pay-v2.ts:60]: @mauricio-sg thanks a lot for this example. I updated and added **type guard**. Take a look please

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/index.ts
Comment (mauricio-sg) [inline packages/core/src/checkout-buttons/strategies/amazon-pay-v2/index.ts:7]: Not sure if mocks should be indexed? 🤔

**Review by mauricio-sg — APPROVED** [review]
Well, it looks good to me :)

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/index.ts
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/amazon-pay-v2/index.ts]: @mauricio-sg thanks, fixed

**Review by serhii-tkachenko — APPROVED** [review]
Overall looks good to me! Nice work @bc-nick 👍

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-config-request-sender.ts
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-config-request-sender.ts]: @serhii-tkachenko thanks. fixed all according to your comments

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/errors/amazon-pay-v2-config-creation-error.ts
Comment (animesh1987) [inline packages/core/src/checkout-buttons/strategies/amazon-pay-v2/errors/amazon-pay-v2-config-creation-error.ts:7]: Have you looked at translations for this?

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-request-sender.ts
Comment (animesh1987) [inline packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-request-sender.ts:22]: What does this endpoint do ?

Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-request-sender.ts:22]: This endpoint returns the required information by cartId (quote) to correctly initialize Amazon when the button is clicked.

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts
Comment (animesh1987) [inline packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts:7]: We should be avoiding the use of `as` keyword and looking at stronger typing.

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts
Comment (animesh1987) [inline packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts:22]: Out of curiosity why don't we do the same as paypal, where buy now part becomes an optional in the options object. cc @serhii-tkachenko

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/errors/amazon-pay-v2-config-creation-error.ts
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/amazon-pay-v2/errors/amazon-pay-v2-config-creation-error.ts:7]: We haven't translations for this message. Should we have?
As far as i understood it is not necessary

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts:7]: Why should we avoid this? It is acceptable use it with type guard

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts:22]: @animesh1987 it's not necessary. I think this message should answer of your question https://github.com/bigcommerce/checkout-sdk-js/pull/1742#discussion_r1065126812

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/errors/amazon-pay-v2-config-creation-error.ts
Comment (bc-nick) [reply packages/core/src/checkout-buttons/strategies/amazon-pay-v2/errors/amazon-pay-v2-config-creation-error.ts:7]: i think it would be better to do it in another PR if you don't mind
this message is most for developers than for customers

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts
Comment (animesh1987) [reply packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts:7]: Because `as` itself even in a typeguard can lead to false assertions. Same as saying a `test as object` which implies test is an object 🤔

File: packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts
Comment (animesh1987) [reply packages/core/src/checkout-buttons/strategies/amazon-pay-v2/amazon-pay-v2-button-options.ts:22]: Thanks for the context 🙇

**Review by animesh1987 — APPROVED** [review]
Nice work @bc-nick 👏

**Pull request merged**
